src/ule.rs

```diff
@@ -0,0 +1,101 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{TinyStr4, TinyStr8, TinyStr16, Error};
+use zerovec::ule::{ULE, AsULE, PlainOldULE};
+use std::mem;
+
+/// This is an unaligned little-endian version of TinyStr.
+///
+/// TinyStr is already endian-agnostic (like str), so the only difference is alignment.
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct UTF8ULE<const N: usize>(PlainOldULE<N>);
+
+macro_rules! impl_str_ule_size {
+    ($size:literal, $tiny:ty, $integer:ty) => {
+        impl From<$tiny> for UTF8ULE<$size> {
+            fn from(s: $tiny) -> Self {
+                // This converts between endiannesses twice: TinyStr::into converts
+                // from little-endian into native, and PlainOldULE::from
+                // converts from native to little again
+                let int: $integer = s.into();
+                UTF8ULE(int.into())
+            }
+        }
+
+        impl AsULE for $tiny {
+            type ULE = UTF8ULE<$size>;
+            #[inline]
+            fn as_unaligned(&self) -> Self::ULE {
+                (*self).into()
+            }
+            #[inline]
+            fn from_unaligned(unaligned: &Self::ULE) -> Self {
+                unsafe {
+                    // This is safe since UTF8ULE guarantees that it comes from
+                    // a valid TinyStr
+
+                    // This converts between endiannesses twice: TinyStr::new_unchecked()
+                    // takes in a native endian integer, which we produce via from_unaligned()
+                    Self::new_unchecked(<$integer>::from_unaligned(&unaligned.0))
```

Comment:
Suggestion: Create a `const unsafe` constructor for TinyStr (in the normal library code) that takes a `&[u8; $size]`; it can use `<$integer>::from_ne_bytes()` in its impl. Then here, pass in the bytes.

Reply:
https://github.com/zbraniecki/tinystr/issues/45